### ISSUE TITLE
Local avro schema with nullable fields fails with `Cannot read properties of null (reading 'schemaFormat')` using `@asyncapi/parser@v2.0.2`

### ISSUE THREAD

**Issue by Gyllsdorff** [asyncapi/parser-js#784]
**TL;DR;**
AVRO schema with nullable fields works with `@asyncapi/parser@v2.0.1` but fails when I use `@asyncapi/parser@v2.0.2`

#### Describe the bug

If I my AVRO schema has a nullable field, e.g. `default: null, type: ["null", "string"]`, then it fails with a unclear error when I try to generate a model or when I validate it by doing `npx -p @asyncapi/cli@v0.47.9 asyncapi validate example.yaml`. It works fine and I can generate a HTML doc if I lock down the `@asyncapi/parser` to version `2.0.1` version by doing `npx -p @asyncapi/cli@v0.47.5 -p @asyncapi/parser@v2.0.1 -p @asyncapi/generator@v1.10.4 asyncapi validate example.yaml`.

Generate command:
> Generator Error: Input is not a corrent AsyncAPI document so it cannot be processed.

Validate command:
> File example.yaml and/or referenced documents have governance issues.

If I remote `default: null` from the schema then it can generate the documentation but the schema will not allow the producer to not set the field.

I debugged the code and found that there is a `TypeError: Cannot read properties of null (reading 'schemaFormat')` error that gets discarded which seems to be triggered by [this rule](https://github.com/asyncapi/parser-js/blob/master/src/ruleset/v2/ruleset.ts#L126)

```
TypeError: Cannot read properties of null (reading 'schemaFormat')
    at Object.$.channels.*.[publish,subscribe].message.oneOf[?(@.schemaFormat === void 0)] (eval at query (/Users/<REDACTED>/.npm/_npx/c7940f4513e4e66f/node_modules/nimma/dist/legacy/cjs/core/index.js:66:71), <anonymous>:191:31)
    at eval (eval at query (/Users/<REDACTED>/.npm/_npx/c7940f4513e4e66f/node_modules/nimma/dist/legacy/cjs/core/index.js:66:71), <anonymous>:641:91)
```

#### How to Reproduce

- 🔴 `npx -p @asyncapi/cli@v0.47.9 asyncapi validate example.yaml`
- 🔴 `npx -p @asyncapi/cli@v0.47.5 -p @asyncapi/parser@v2.0.2 -p @asyncapi/generator@v1.10.4 asyncapi validate example.yaml`
- 🟢 `npx -p @asyncapi/cli@v0.47.5 -p @asyncapi/parser@v2.0.1 -p @asyncapi/generator@v1.10.4 asyncapi validate example.yaml`

```
> npx -p @asyncapi/cli@v0.47.9 asyncapi validate example.yaml

File example.yaml and/or referenced documents have governance issues.

> npx -p @asyncapi/cli@v0.47.5 -p @asyncapi/parser@v2.0.2 -p @asyncapi/generator@v1.10.4 asyncapi validate example.yaml

File example.yaml and/or referenced documents have governance issues.

> npx -p @asyncapi/cli@v0.47.5 -p @asyncapi/parser@v2.0.1 -p @asyncapi/generator@v1.10.4 asyncapi validate example.yaml

File kafka-topics.yaml is valid! File kafka-topics.yaml and referenced documents don't have governance issues.
```

<details>

<summary>AsyncAPI file:</summary>

```yaml
asyncapi: 2.6.0
id: https://example.com
tags: []
info:
  title: example
  version: "1.0.0"
  license:
    name: No License
  contact:
    name: Foo
    email: foo@example.com
  description: >
    Short description

channels:
  public.ad-moderation.ad:
    publish:
      operationId: test1
      message:
        messageId: test2
        schemaFormat: "application/vnd.apache.avro;version=1.9.0"
        payload:
          $ref: ./example1.avsc.yaml
```

</details>

<details>

<summary>AVRO schema:</summary>

```yaml
name: Command
type: record
fields:
  - name: foo
    # Works if you add a comment away `default: null`.
    default: null
    type:
      - "null"
      - string
```

</details>

#### Expected behavior

`asyncapi validate` should return OK and `asyncapi generate [...]` should generate a HTML documentation.

**Comment by jonaslagoni** [comment]
cc @chrispatmore

**Comment by chrispatmore** [comment]
Yep, this look like it was me will fix

**Comment by chrispatmore** [comment]
Hey sorry to bother you @Gyllsdorff I am struggling to pin this down. the fact that it varies based on `default: null` is even more baffling. How did you do debug the code to find the above error message w.r.t `Cannot read properties of null (reading 'schemaFormat')`

**Comment by chrispatmore** [comment]
Alright so this is a real brain teaser. As I can understand it, `spectral` uses the `nimma` package to evaluate the jsonpath rules. `nimma` seems to work by first parsing and checking all the jsonpaths are valid (which they are) then traversing the whole json object (in this case the async api) and for each node and each field in the node checking if it meets the jsonpath.

The problematic rule / null combo in question that `nimma` generates is
```
"$.channels.*.[publish,subscribe].message.oneOf[?(@.schemaFormat === void 0)]": function (scope) {
    if (scope.depth < 5) return;
    if (scope.path[0] !== "channels") return;
    if (!(scope.sandbox.value.schemaFormat === void 0)) return;
    if (scope.path[scope.depth - 1] !== "oneOf") return;
    if (scope.path[scope.depth - 2] !== "message") return;
    if (scope.path[scope.depth - 3] !== "publish" && scope.path[scope.depth - 3] !== "subscribe") return;
    scope.emit("$.channels.*.[publish,subscribe].message.oneOf[?(@.schemaFormat === void 0)]", 0, false);
  },
```

And it is trying to apply this against the path `channels.public.ad-moderation.ad.publish.message.payload.fields.0.default` for which the value is `null`. e.g. `scope.sandbox.value` is null. At which point you can see where the `npe` comes from. `scope.sandbox.value.schemaFormat` which is applied as the third "rule". 

what I don't understand is why `nimma` is checking schema format so deep `scope.depth` here is `8` rather than just checking at depth 5, or indeed why it applies the rules in that order.

I will attempt to find a variation on the rule that only checks for schemaFormat at top depth so we don't hit this

**Comment by chrispatmore** [comment]
Got a fix: https://github.com/asyncapi/parser-js/pull/786
Tried a bunch of rule variations and the only one that gave the correct output was to explicitly check that the value was not null. 
Which gets nimma to generate a rule like:
```
"$.channels.*.[publish,subscribe].message.oneOf[?(!@null && @.schemaFormat === void 0)]": function (scope) {
    if (scope.depth < 5) return;
    if (scope.path[0] !== "channels") return;
    if (!(!(scope.sandbox.value === null) && scope.sandbox.value.schemaFormat === void 0)) return;
    if (scope.path[scope.depth - 1] !== "oneOf") return;
    if (scope.path[scope.depth - 2] !== "message") return;
    if (scope.path[scope.depth - 3] !== "publish" && scope.path[scope.depth - 3] !== "subscribe") return;
    scope.emit("$.channels.*.[publish,subscribe].message.oneOf[?(!@null && @.schemaFormat === void 0)]", 0, false);
  }
```
which defends `scope.sandbox.value.schemaFormat ` with a null check

in the process tested by generating the rules directly with `nimma`:
```
import Nimma from 'nimma';
import fs from 'fs';

const n = new Nimma([
    '$.channels.*.[publish,subscribe].message.oneOf[?(@.schemaFormat === void 0)]',
    '$.channels.*.[publish,subscribe].message.oneOf[?(@object !== null && @.schemaFormat === void 0)]',
    '$.channels.*.[publish,subscribe].message.oneOf[?(!@null && @.schemaFormat === void 0)]',
]);

fs.writeFile('./nimma-code.mjs', n.sourceCode, err => {
    if (err) {
        console.error(err);
    }
});
```
To see what code was actually going to run. probably worth remembering

**Comment by Gyllsdorff** [comment]
Nice work!
